devito/finite_difference.py

```diff
@@ -221,3 +221,72 @@ def second_cross_derivative(function, dims, order):
     """
     first = second_derivative(function, dim=dims[0], width=order)
     return second_derivative(first, dim=dims[1], order=order)
+
+
+def staggered_diff(f, dim, order, stagger=centered, theta=0, phi=0):
+    """
+    Utility function to generate staggered derivatives
+    :param f: function objects, eg. `f(x, y)` or `g(t, x, y, z)`.
+    :param dims: symbol defining the dimension wrt. which
+       to differentiate, eg. `x`, `y`, `z` or `t`.
+    :param order: Order of the coefficient discretization and thus
+                  the width of the resulting stencil expression.
+    :param stagger: Shift for the FD, `left`, `right` or `centered`
+    :param theta: Dip (or polar) angle for rotated FD
+    :param phi: Azimuth angle for rotated FD
+    """
+    ndim = len(f.space_dimensions)
+    off = dict([(d, 0) for d in f.space_dimensions])
+    if stagger == left:
+        off[dim] = -.5
+    elif stagger == right:
+        off[dim] = .5
+    else:
+        off[dim] = 0
+
+    if theta == 0 and phi == 0:
+        diff = dim.spacing
+        idx = list(set([(dim + int(i+.5+off[dim])*diff)
+                        for i in range(-int(order / 2), int(order / 2))]))
+        if int(order / 2) == 1:
+            idx = [dim - diff, dim]
+        return f.diff(dim).as_finite_difference(idx, x0=dim + off[dim]*dim.spacing)
+    else:
+        if ndim < 2 or ndim > 3:
+            error("Only 2D and 3D supports rotated finite difference,"
+                  " %d dimensions as input" % ndim)
+            raise ValueError("Non supported number of dimensions")
+
+        x = f.space_dimensions[0]
+        z = f.space_dimensions[-1]
+        idxx = list(set([(x + int(i+.5+off[x])*x.spacing)
+                         for i in range(-int(order / 2), int(order / 2))]))
+        dx = f.diff(x).as_finite_difference(idxx, x0=x + off[x]*x.spacing)
+        dx = dx.evalf(9)
```

Comment:
this is a hidden magic number, and it should be configurable somehow. At least put it at the top of the file as "PRECISION=9" (why 9? why not more?) or with some other names...
Also, are we sure we're not affected by the same issue in the other FD functions?

Reply:
I'm investigating other function.
9 is not "magic, single precision is roughly 6 to 9 digits so top end of single precision, double would be 15 butit's where it breaks.

Reply:
the single precision format should be *at least* 9 digits. But yeah, I called it "magic number" because in jargon that's how people refer to unnamed constants popping up in the code

Reply:
>the single precision format should be at least 9 digits

It's at most 9 decimal digits  in single precision

But yes it's a bit magic for now while I investigate it further